### PR TITLE
changed groupby function to accept multiple attributes #1839

### LINKED ISSUE #1839: groupby should support multiple attrs
### Expected behaviour

`ag.groupby(('segids', 'resnames'))` gives a dictionary with the key being a combination of `segid` and `chainid`. For example `ag.groupby(('segid', 'resnames'))['A']['TYR']` would work. This is useful where one has to iterate over combinations of attributes.

### Actual behaviour

Only one attr is supported.

### Code

Changes need to be made here
https://github.com/MDAnalysis/mdanalysis/blob/63b70f068ce4eb4583ada2efbe4cb0ed1ee67ba1/package/MDAnalysis/core/groups.py#L1042

### Currently version of MDAnalysis:
(run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
0.18.0-dev0

### PR DESCRIPTION
**Pull request by davidercruz**
Fixes #1839 

Changes made in this Pull Request:
 - Changes to the groupby function now allow multiple attributes

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/core/groups.py [hunk 1]
```diff
@@ -1039,32 +1039,48 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+            12.011: <AtomGroup with 57 atoms>,
+            14.007: <AtomGroup with 19 atoms>,
+            15.999: <AtomGroup with 19 atoms>},
+           'ARG': {1.008: <AtomGroup with 169 atoms>,
+           ...}
+          
+          >>> ag.groupby('resnames', 'masses')['ALA'][15.999]
+           <AtomGroup with 19 atoms>
+        
+        .. versionadded::
```
Comment (jbarnoud) [inline package/MDAnalysis/core/groups.py]: The version number disappeared here. Also, you should add a line indicating that the function changed in version 0.18.0:

```
.. versionchanged:: 0.18.0

    The function accepts multiple attributes.
```

**Comment by orbeckst** [comment]
By the way, test failure (e.g., https://travis-ci.org/MDAnalysis/mdanalysis/jobs/358144131#L2019) from ../../../miniconda/envs/test/lib/python3.6/site-packages/scipy/optimize/minpack.py is very likely not related to this PR but has been coming up more and more recently. Possibly restarting the tests will help (or when you commit a commit with the tests the CI will run again).

@davidercruz In the future, please check the output from Travis (follow the "Details" link from *continuous-integration/travis-ci*) and if it does not look like failures related to your code you can just state that in a comment. A core-dev might just rerun the tests or to decide to merge nevertheless but it helps to already being alerted to why the tests are not passing – some of us don't bother reviewing PRs that don't have a green check mark for the tests.

File: package/MDAnalysis/core/groups.py [hunk 2]
```diff
@@ -1039,32 +1039,48 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+            12.011: <AtomGroup with 57 atoms>,
+            14.007: <AtomGroup with 19 atoms>,
+            15.999: <AtomGroup with 19 atoms>},
+           'ARG': {1.008: <AtomGroup with 169 atoms>,
+           ...}
+          
+          >>> ag.groupby('resnames', 'masses')['ALA'][15.999]
+           <AtomGroup with 19 atoms>
+        
+        .. versionadded:: 0.18.0
```
Comment (jbarnoud) [inline package/MDAnalysis/core/groups.py]: The function has not been added in version 0.18.0, it has been added in version 0.16.0 and changed in 0.18.0.

File: package/MDAnalysis/core/groups.py [hunk 3]
```diff
@@ -1039,32 +1039,48 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+            else:
+                res[i] = self[ta == i].groupby(*topattrs[1:])
+        return res
 
-          >>> ag.groupby('masses')
-          {12.010999999999999: <AtomGroup with 462 atoms>,
-          14.007: <AtomGroup with 116 atoms>,
-          15.999000000000001: <AtomGroup with 134 atoms>}
-
-        .. versionadded:: 0.16.0
```
Comment (jbarnoud) [inline package/MDAnalysis/core/groups.py]: This is the line that should have stayed.

File: package/MDAnalysis/core/groups.py [hunk 4]
```diff
@@ -1039,32 +1039,49 @@ def wrap(self, compound="atoms", center="com", box=None):
...
         Returns
         -------
         dict
-            Unique values of the topology attribute as keys, Groups as values.
+            Unique values of the multiple combinations of topology attributes 
+            as keys, Groups as values.
 
         Example
         -------
-        To group atoms with the same mass together::
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: Leave in the example with a single argument.

File: testsuite/MDAnalysisTests/core/test_groups.py [hunk 5]
```diff
@@ -576,7 +576,7 @@ class TestGroupBy(object):
     # tests for the method 'groupby'
     @pytest.fixture()
     def u(self):
-        return make_Universe(('types', 'charges', 'resids'))
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/core/test_groups.py]: Why did you change it from "types" to "segids"? Just aesthetics?

Comment (davidercruz) [reply testsuite/MDAnalysisTests/core/test_groups.py]: I changed it because "type" and "charges" match one to one, for example TypeA matches only charge -1.5 . So, to keep the structure of the tests consistent, I changed "types" to "segids"

Comment (jbarnoud) [reply testsuite/MDAnalysisTests/core/test_groups.py]: If it matches 1 to 1, then there are bugs you will not catch. You want the attributes in your test to match as little as possible to each level of grouping is different from the other.

Comment (jbarnoud) [reply testsuite/MDAnalysisTests/core/test_groups.py]: I misread, so yes changid *to* 'segid' was the right thing to do.

File: package/MDAnalysis/core/groups.py [hunk 6]
```diff
@@ -1039,32 +1039,58 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+        To group atoms with the same mass together::     
+                
+        >>> ag.groupby('masses')
+        {12.010999999999999: <AtomGroup with 462 atoms>,
+         14.007: <AtomGroup with 116 atoms>,
+         15.999000000000001: <AtomGroup with 134 atoms>}        
+
+        To group atoms with the same residue name and mass together::
+
+          >>> ag.groupby('resnames', 'masses')
```
Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: About the API. It might be better to follow the pandas/datreant examples and use an explicit list for multiple arguments. Then our groupby API is as close as possible to others minimizing surprises for the user.

```python
# MDAnalysis
ag.groupby(['masses', 'charges'])
# pandas
df.groupby(['columnname', 'othercolumn'])
# datreant
bundle.categories.groupby(['category1', 'categorie2'])
```

**Comment by davidercruz** [comment]
theres a typo in my last commit. Instead of functions it shoud read arguments.

File: package/MDAnalysis/core/groups.py [hunk 7]
```diff
@@ -1039,7 +1039,7 @@ def wrap(self, compound="atoms", center="com", box=None):
             if not all(s == 0.0):
                 o.atoms.translate(s)
 
-    def groupby(self, *topattrs):
+    def groupby(self, topattrs):
```
Comment (jbarnoud) [inline package/MDAnalysis/core/groups.py]: Why did you change that? Having `topattrs` as a star argument was allowing not to break existing code.

Comment (davidercruz) [reply package/MDAnalysis/core/groups.py]: The problem was that, while it still worked when using a single argument, when providing a list of two or more arguments, topattrs becomes a tuple where the first element is a list, and the second is empty. I'm sure there's a better way of doing this

Comment (kain88-de) [reply package/MDAnalysis/core/groups.py]: See https://github.com/MDAnalysis/mdanalysis/pull/1846#discussion_r177389219

But the current version shouldn't break existing code

Comment (jbarnoud) [reply package/MDAnalysis/core/groups.py]: OK I missed @kain88-de's comment.

File: package/MDAnalysis/core/groups.py [hunk 8]
```diff
@@ -1039,32 +1039,61 @@ def wrap(self, compound="atoms", center="com", box=None):
             if not all(s == 0.0):
                 o.atoms.translate(s)
 
-    def groupby(self, topattr):
+    def groupby(self, topattrs):
         """Group together items in this group according to values of *topattr*
 
         Parameters
         ----------
         topattr: str
```
[the remark this answers is not in the document]
Comment (davidercruz) [reply package/MDAnalysis/core/groups.py]: could it be
topattrs: str / list for single / multiple atributes
?

**Comment by kain88-de** [comment]
Before this is merged can we change the return to a single dictionary?

**Comment by davidercruz** [comment]
@kain88-de The current function returns a dictionary, a dictionary of dictionaries when using two attributes. 
What do you have in mind?

**Comment by kain88-de** [comment]
Here with examples from other tools using `groupby` 

```python
# pandas (they have a groupby special object)
gb = df.groupby(['col1', 'col2'])
group = gb.get_group(('var1', 'var2'))
# datreant
gb = treant.categories.groupby(['cat1', 'cat2'])
gb[('var1', 'var2')]
# MDAnalysis
gb = ag.groupby(['attr1', 'attr2'])
gb[('var', 'var')]
```

I'm sorry for not being clear about this on the issue.   I just think an API that is similar to other tools would be more intuitive to use.

**Comment by orbeckst** [comment]
I approved the commit to signal that I am happy with the progress that @davidercruz made. 

However, I don't want to merge right now without understanding if we intend to break old code. If we intend to change the API for groupby() then we should have at least a deprecation period where we allow single argument but issue a deprecation warning. If we want to go down this route then I suggest we 
1. merge the PR as is
2. raise a new issue to add the old behavior as deprecated

File: package/CHANGELOG [hunk 9]
```diff
@@ -19,6 +19,9 @@ mm/dd/18 richardjgowers, palnabarun, bieniekmateusz, kain88-de, orbeckst,
   * 0.17.1
 
 Enhancements
+  * Can now pass multiple attributes as a list to groupby function.
```
Comment (orbeckst) [inline package/CHANGELOG]: @davidercruz also add your GitHub hand *davidercruz* to the list of authors above (after ayshsuhane)

**Comment by richardjgowers** [comment]
@orbeckst I think both `gb(['string'])` and `gb['string']` will work currently so old behaviour isn't broken

I think Max doesn't want nested dictionaries as the return but instead a single dictionary with long keys

**Comment by davidercruz** [comment]
@orbeckst yes, I need this for my submission.

I find the current solution is intuitive for iterating over the dictionary.

@kain88-de from what I understand, you propose the output to be a dict where the key is a list. I can implement this, but I can't think of a clean way to do this with the existing code, right now. I have to look into it.

**Comment by orbeckst** [comment]
Ok @richardjgowers – it also turns out I misunderstood how the `*args` were processed; we actually do check if it is a `str`. I will amend my previous comment...

**Comment by orbeckst** [comment]
@davidercruz I think you have to "flatten" your nested dict:
```python
>>> tt = {'a': {'1': 'a1', '2': 'a2'}, 'b': {'1': 'b1', '2': 'b2'}}
>>> ff = {}
... for k1, v1 in tt.items():
...     for k2, v2 in v1.items():
...        ff[(k1, k2)] = v2
...
>>> ff
{('a', '1'): 'a1', ('a', '2'): 'a2', ('b', '1'): 'b1', ('b', '2'): 'b2'}
```

**Comment by orbeckst** [comment]
simpler:
```python
ff= {(k1, k2): v2 for k2,v2 in v1.items() for k1,v1 in tt.items()}
```

File: package/MDAnalysis/core/groups.py [hunk 10]
```diff
@@ -1039,32 +1039,63 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+                
+        >>> ag.groupby('masses')
+        {12.010999999999999: <AtomGroup with 462 atoms>,
+         14.007: <AtomGroup with 116 atoms>,
+         15.999000000000001: <AtomGroup with 134 atoms>}        
+
+        To group atoms with the same residue name and mass together::
+
+          >>> ag.groupby(['resnames', 'masses'])
+          {'ALA': {1.008: <AtomGroup with 95 atoms>,
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: Change output to reflect the "flat" dict.

File: package/MDAnalysis/core/groups.py [hunk 11]
```diff
@@ -1039,32 +1039,63 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+            ta = getattr(self, topattrs)
+            return {i: self[ta == i] for i in set(ta)}
+        else:
+            ta=getattr(self,topattrs[0])
+            for i in set(ta):
+                if len(topattrs) == 1:
+                    res[i] = self[ta == i]
+                else:
+                    res[i] = self[ta == i].groupby(topattrs[1:])
+            return res
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: Return a dict with keys as tuples; see perhaps https://stackoverflow.com/a/6027615 for a general `flatten()` function that can be adapted to produce key tuples.

Comment (orbeckst) [reply package/MDAnalysis/core/groups.py]: See https://github.com/MDAnalysis/mdanalysis/pull/1846#issuecomment-376699783 for a general `flatten()` that creates tuple keys.

If that version of `flatten()` works then you can put it into `lib.utils.flatten_dict()` or similar. Add as a test case the example that I used.

Comment (davidercruz) [reply package/MDAnalysis/core/groups.py]: @orbeckst Thank you for your help.

I got this function working to flatten the output of groupby
```python
    import collections
    def flatten_dict(d, parent_key=''):
        items = []
        for k, v in d.items():
            new_key = (parent_key, k) if parent_key else k
            if isinstance(v, collections.MutableMapping):
                 items.extend(flatten_dict(v, new_key).items())
            else:
            items.append((new_key, v))
            return dict(items)
```

Comment (davidercruz) [reply package/MDAnalysis/core/groups.py]: @orbeckst Somehow I missed the edit to your comment https://github.com/MDAnalysis/mdanalysis/pull/1846#issuecomment-376699783

Comment (orbeckst) [reply package/MDAnalysis/core/groups.py]: submit a PR and ping me and I will review.

Comment (orbeckst) [reply package/MDAnalysis/core/groups.py]: ... "submit a PR" – I meant, just push a new commit with the changes.

Comment (davidercruz) [reply package/MDAnalysis/core/groups.py]: @orbeckst I tried your code, but it fails on this line:
``` new_key = parent_key+tuple(k)
TypeError: 'numpy.float32' object is not iterable
```
while using (k, ) instead works fine

Comment (orbeckst) [reply package/MDAnalysis/core/groups.py]: Use what works :-)

**Comment by orbeckst** [comment]
@davidercruz 
cc @jbarnoud @richardjgowers )

> yes, I need this for my submission.

Please do submit and finalize your GSOC application even if merging this PR is slowed down.

File: package/MDAnalysis/core/groups.py [hunk 12]
```diff
@@ -1039,32 +1039,63 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+           <AtomGroup with 19 atoms>
+        
         .. versionadded:: 0.16.0
+        .. versionchanged:: 0.18.0 The function accepts multiple attributes
         """
-        ta = getattr(self, topattr)
-        return {i: self[ta == i] for i in set(ta)}
+        
+        res = {}
+        if type(topattrs) == str:
```
Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: For python2/3 compatibility at the top please to `from six import string_types` and replace this line with `if isinstance(topattrs, string_types)`

File: testsuite/MDAnalysisTests/core/test_groups.py [hunk 13]
```diff
@@ -603,6 +603,44 @@ def test_groupby_int(self, u):
...
+                assert (ref, subref) in gb.keys()
+                a = gb[(ref, subref)]
+                assert len(a) == 5
+                assert all(a.charges == ref)
+                assert all(a.segids  == subref)
+
+    def test_groupby_int_float(self, u):
+        gb = u.atoms.groupby(['resids', 'charges'])
+
+        uplim=len(gb)/5+1
```
Comment (kain88-de) [inline testsuite/MDAnalysisTests/core/test_groups.py]: This value will be a `float` in python3. This is because any division in python3 produces a float. You will need an explicit int cast.

File: package/MDAnalysis/core/groups.py (hunk 12 above)
Comment (kain88-de) [reply package/MDAnalysis/core/groups.py]: Tests should also check for `b'masses'`, 'u'masses'` as well to check for all possible string types in python.

Comment (kain88-de) [reply package/MDAnalysis/core/groups.py]: These are different types of strings and they behave slightly different in python2 and 3. You can read this `b'ab'` as bytes and `u'ab'` as a Unicode string. The behavior is different because in Python3 the default strings are Unicode. We need to test against them.

**Comment by jbarnoud** [comment]
The best way to test may be something in the line of `isinstance(topattr, six.string_types)`.

**Comment by davidercruz** [comment]
@kain88-de  @orbeckst 

Since string_types is already imported for six in groups.py it can be simply:
```python
if isinstance(topattrs, string_types):
```
right?

**Comment by kain88-de** [comment]
if `string_types` is already imported you can simply use it yes.

File: testsuite/MDAnalysisTests/core/test_groups.py [hunk 14]
```diff
@@ -588,13 +588,13 @@ def test_groupby_float(self, u):
             assert len(g) == 25
 
     def test_groupby_string(self, u):
-        gb = u.atoms.groupby('types')
+        gb = u.atoms.groupby('segids')
```
Comment (kain88-de) [inline testsuite/MDAnalysisTests/core/test_groups.py]: you can parametrize this test on string types.

```python
@pytest.mark.parametrize('string', ['segids', b'segid', u'segids'])
def test_groupby_string(self, u, string):
    gb = u.atoms.groupby(string)
    ...
```

File: package/MDAnalysis/core/groups.py [hunk 15]
```diff
@@ -1039,32 +1039,63 @@ def wrap(self, compound="atoms", center="com", box=None):
...
         -------
-        To group atoms with the same mass together::
-
-          >>> ag.groupby('masses')
-          {12.010999999999999: <AtomGroup with 462 atoms>,
-          14.007: <AtomGroup with 116 atoms>,
-          15.999000000000001: <AtomGroup with 134 atoms>}
-
+       
+
```
Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: I think sphinx complains about the two empty lines. If you remove them the doc test should pass.

File: testsuite/MDAnalysisTests/core/test_groups.py (hunk 14 above)
Comment (davidercruz) [reply testsuite/MDAnalysisTests/core/test_groups.py]: @kain88-de thhe test on string types fails for b'segids' on python3 with a 

```
 TypeError: getattr(): attribute name must be string
```
Using the decode method on toppattrs in the groupby function would solve this issue. Which encoding should I use: ascii or utf-8 ?

**Comment by davidercruz** [comment]
grouby test was failing for b'segids'. I implemented a check for this kind of strings and decoded them to utf-8 in the groupby function. Now the test does not fail.

File: package/MDAnalysis/core/groups.py [hunk 16]
```diff
@@ -1039,32 +1039,67 @@ def wrap(self, compound="atoms", center="com", box=None):
...
+        .. versionadded:: 0.16.0
+        .. versionchanged:: 0.18.0 The function accepts multiple attributes
+        """
+        
+        res = dict()
+        
+        if isinstance(topattrs, (string_types, bytes)):
+            topattrs = (topattrs,)
+    
+        if isinstance(topattrs[0], bytes):
```
Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: So here I learned something new myself. bytes isn't a string type in Python 3, I should have known that. Could be I was a bit to conservative with testing. We can rather just test unicode string and not raw bytes. But let's see how this goes and if it fails for anyone in the future. Should be a rare case.

**Comment by kain88-de** [comment]
@davidercruz your own added tests fail now. If this is because of the bytes issue I would remove that that again and the check in the function. It's not worth it and a explicit decoding is OK.  I would leave the Unicode test with `u'segids'` though.

**Comment by davidercruz** [comment]
@kain88-de I think it's good now. In the previous commit the keys of the output dict where tuples even for single attributes. Tested it for both Python 2 and 3 and tests are OK

**Comment by orbeckst** [comment]
It's again minpack test failures #1848, so I think this is ready to be merged.

However, I'd like to clean up the history a little bit. There are quite a few commits that are small changes. I would either all squash it all into a single commit (easy) or have it in 2-4 commits, along the lines of
- `flatten_dict()` with its tests
- new groupby implementation
- tests (or keep with groupby)
- meta data updates (or keep it with the code commits)

@davidercruz what is your opinion?

**Comment by davidercruz** [comment]
@orbeckst Yes, there are a lot. I think a single commit would be fine. I am new to git - is the commit squash made by me?

**Comment by orbeckst** [comment]
No, you don't have to do the squash – this can be done by whoever merges. If you're ok with it, then we'll do a **squash and merge**.

Rewriting history with [`git rebase -i`](https://git-scm.com/docs/git-rebase) is a bit more tricky and although you probably eventually want to learn how to do it, it is not the first thing I recommend to someone learning git because there's the potential for data loss.

**Comment by davidercruz** [comment]
@orbeckst 
OK. Thank you, I will look into it.

I learned a lot with this PR. @orbeckst @jbarnoud @kain88-de  you have been all very helpful. Thanks :-)

**Comment by kain88-de** [comment]
@jbarnoud  or @richardjgowers if you don't have any more issues this it good to go.

**Comment by jbarnoud** [comment]
Congrats @davidercruz for your first merged PR :smiley:
